### PR TITLE
[codex] Adapt skill bundles into Reborn skill context

### PR DESCRIPTION
**Pull request by serrrfirat**
## Summary
- add `SkillBundleContextSource`, a `HostSkillContextSource` adapter over `SkillBundleSource`
- read raw `SKILL.md` only for visible bundles with present trust/visibility metadata
- preserve deterministic ordering via bundle descriptor ordering keys
- keep installed skill prompts out of model-visible snippets while allowing trusted prompts through
- fail closed on missing policy metadata without reading bundle content

## Context
Stacked on #3847. This is PR 5 from #3473 first-party skills extension lane: bundle source -> runtime skill-context candidate adapter. Runtime composition wiring remains a follow-up once the source stack lands.

## Validation
- `cargo fmt --check -p ironclaw_loop_support`
- `git diff --check`
- `cargo test -p ironclaw_loop_support skill_bundle_context`
- `cargo test -p ironclaw_loop_support`
- `cargo clippy -p ironclaw_loop_support --all-targets -- -D warnings`

### CODE REVIEW

File: crates/ironclaw_loop_support/src/skill_bundle_context_source.rs [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+            if trust.is_none() || visibility != Some(SkillVisibility::Visible) {
+                candidates.push(
+                    HostSkillContextCandidate::unavailable(trust, visibility)
+                        .with_ordering_key(ordering_key),
+                );
+                continue;
+            }
+
+            let skill_md = self
```
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/skill_bundle_context_source.rs]: **Performance:** Sequential SKILL.md reads. For N visible bundles, latency = Σ(read_time_i). Consider fanning out reads with `join_all` / `try_join_all`.

**Tests:** Non-UTF8 parse path at line 85 is untested. Also missing: PermissionDenied read error test, ContentTooLarge mapping test. Consider adding these.

**Docs:** Public methods `new()` and `bundle_source()` lack doc-comments.

**Review by serrrfirat — COMMENTED** [review]
## PR Review Summary
**PR:** nearai/ironclaw#3848 — Adapt skill bundles into Reborn skill context
**SHA:** 5135fc38c3a6dee2394129bbc0393545c2a5fa85
**Findings:** 6 total (3 Medium, 2 Low, 1 Low)

| Category | Critical | High | Medium | Low |
|---|---|---|---|---|
| Security | 0 | 0 | 0 | 0 |
| Bugs | 0 | 0 | 0 | 0 |
| Performance | 0 | 0 | 1 | 1 |
| Tests | 0 | 0 | 3 | 0 |
| Conventions | 0 | 0 | 1 | 0 |

**Intent:** Add SkillBundleContextSource adapter to expose skill bundles as model-visible context candidates, with trust/policy gating and deterministic ordering. Step 5 in the #3473 extension chain.

**Verdict:** COMMENT — clean security and bugs. Test coverage gaps and a performance concern should be addressed.

---

## Performance

### Medium: Sequential SKILL.md reads
`load_skill_context_candidates` calls `read_skill_bundle_file()` serially (line 79-81). For B visible bundles: latency = Σ(read_i). Since `build_skill_instruction_snippets()` runs per loop tick, N bundles × sequential reads adds up.

**Suggestion:** Fan out reads with `join_all` / `try_join_all` on bounded tasks, then reconstruct in descriptor order.

### Low: Redundant SKILL.md parse
`read_skill_bundle_file` calls `validate_bundle_manifest` on every read. `list_skill_bundles` already parses and validates the manifest during discovery. Each visible bundle gets SKILL.md re-parsed twice.

## Tests

### Medium: Non-UTF8 skill content not tested
Line 85 calls `String::from_utf8(skill_md).map_err(...)`. No test provides invalid UTF-8 to exercise the `ParseFailed` path. Safe defaults but untested.

### Medium: PermissionDenied read error not tested
`PermissionDenied → SourceUnavailable` error mapping at line 101 is never exercised. The mock only ever returns `FileNotFound`.

### Medium: ContentTooLarge not tested
`ContentTooLarge → ContextBudgetExceeded` mapping at lines 93-96 is dead code without a test. Should verify the adapter bails out when a bundle exceeds budget.

## Conventions

### Medium: Missing doc-comments on pub methods
`new()` (line 28) and `bundle_source()` (line 32) lack `///` documentation. The crate convention (e.g., `cancellation_port.rs`) documents all pub surfaces. Peer type `HostSkillContextSource` and `HostSkillContextCandidate` in the same module are fully documented.

---
⚠️ Review submitted by code-review skill v1

**Review by serrrfirat — COMMENTED** [review]
Multi-agent code review for PR #3848 at `5135fc38c3a6dee2394129bbc0393545c2a5fa85`.

Reviewers: security, bugs, performance/concurrency, tests, conventions.

Findings: 4 total
- Medium: 2
- Low: 2
- Clean reviewers: bugs, performance/concurrency, conventions

The main issue is that the adapter starts reading bundle content before it has proven the full descriptor set has complete policy metadata. The remaining findings are focused test coverage gaps around that boundary and descriptor ordering behavior.

File: crates/ironclaw_loop_support/src/skill_bundle_context_source.rs (hunk 1 above)
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/skill_bundle_context_source.rs]: **Low: Policy gaps are detected after earlier bundle content reads**

The loop reads `SKILL.md` as soon as it sees a visible/trusted descriptor, but missing trust/visibility metadata for a later descriptor is only surfaced later when `build_skill_run_snapshot` walks the returned candidates. That means a corrupt descriptor set can fail closed only after earlier raw prompt content has already been read, which weakens the stated boundary of failing closed on missing policy metadata without reading bundle content.

Consider pre-scanning descriptors for any missing trust or visibility metadata before calling `read_skill_bundle_file`, or returning a policy error/all-unavailable candidate set before any content read occurs.

File: crates/ironclaw_loop_support/src/skill_bundle_context_source.rs [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+            .map_err(skill_bundle_source_error_to_context_error)?;
+        sort_skill_bundle_descriptors(&mut descriptors);
+
+        let mut candidates = Vec::with_capacity(descriptors.len());
+        for descriptor in descriptors {
+            let trust = descriptor.trust().cloned();
+            let visibility = descriptor.visibility().copied();
+            let ordering_key = descriptor.id().to_string();
+
+            if trust.is_none() || visibility != Some(SkillVisibility::Visible) {
```
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/skill_bundle_context_source.rs]: **Medium: Missing test for absent visibility metadata fail-closed path**

The tests cover missing trust and hidden/denied visibility, but not the separate case where trust is present and visibility is `None`. Since this branch is the adapter's fail-closed gate before content reads, add a test such as `adapter_fails_closed_when_visibility_metadata_is_missing_without_reads` that asserts `VisibilityDataMissing` and no source reads.

File: crates/ironclaw_loop_support/src/skill_bundle_context_source.rs [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+            .list_skill_bundles(run_context)
+            .await
+            .map_err(skill_bundle_source_error_to_context_error)?;
+        sort_skill_bundle_descriptors(&mut descriptors);
+
+        let mut candidates = Vec::with_capacity(descriptors.len());
+        for descriptor in descriptors {
+            let trust = descriptor.trust().cloned();
+            let visibility = descriptor.visibility().copied();
+            let ordering_key = descriptor.id().to_string();
```
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/skill_bundle_context_source.rs]: **Medium: Descriptor path ordering is not covered by adapter tests**

The PR goal says ordering should preserve bundle descriptor ordering keys, and `SkillBundleDescriptor::ordering_key` includes the manifest path. The current adapter test only varies source/name pairs, so it would not catch a regression that drops the descriptor path from candidate ordering. Add a case like `adapter_preserves_descriptor_path_in_candidate_ordering` with same source/name descriptors using `SKILL.md` and a nested manifest path.

File: crates/ironclaw_loop_support/src/skill_bundle_context_source.rs [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+#[async_trait]
+impl<S> HostSkillContextSource for SkillBundleContextSource<S>
+where
+    S: SkillBundleSource + ?Sized,
+{
+    async fn load_skill_context_candidates(
+        &self,
+        run_context: &LoopRunContext,
+    ) -> Result<Vec<HostSkillContextCandidate>, HostSkillContextBuildError> {
+        let mut descriptors = self
```
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/skill_bundle_context_source.rs:66]: **Low: Adapter source error paths lack direct tests**

The new adapter maps list/read `SkillBundleSourceError` values and invalid UTF-8 into `HostSkillContextBuildError`, but the tests only exercise successful list/read calls and policy skips. Add a targeted test such as `adapter_maps_bundle_source_errors_to_context_errors` covering list failure, visible bundle read failure, and invalid UTF-8 content.

**Review by serrrfirat — COMMENTED** [review]
## Review Summary

| Category | Findings |
|----------|----------|
| Security | ✓ 0 issues |
| Bugs | ✓ 0 issues |
| Performance | ✓ 0 issues |
| Tests | ✓ 5 tests cover all major paths |
| Conventions | ✓ No violations |

**Overall:** Clean implementation. Trust/visibility gating works correctly, error handling exhaustive, tests comprehensive.

## Intent

Adapt skill bundles into Reborn skill context via a `HostSkillContextSource` adapter that filters by visibility and trust policy (PR 5 from #3473 stack).

## Detailed Analysis

### Security ✓
- Trust check BEFORE file read
- Visibility filter only allows `Visible` bundles  
- Fail-closed on missing trust metadata
- UTF-8 validation prevents injection
- No hardcoded secrets
- `Arc` for thread safety
- No `unsafe` blocks

### Bugs ✓
- Error mapping exhaustive (9/9 variants)
- Sort before iteration (deterministic)
- Pre-allocation via `Vec::with_capacity`

### Performance ✓
- Async I/O for file reads
- Early `continue` for non-visible bundles
- Sequential reads acceptable (no N+1)

### Tests ✓
5 tests covering:
1. visible+trusted happy path
2. installed (non-trusted) filtration
3. hidden/denied bypass
4. missing trust fail-closed
5. deterministic ordering

### Conventions ✓
- Doc comments on public items
- Where clauses for generics
- `#[async_trait]` marker
- `Result`-based error handling
- snake_case naming

## Recommendation

**No changes required.** Implementation is production-ready.

**Comment by serrrfirat** [comment]
Addressed the review findings in `e1a6ba55`:

- Added a fail-closed descriptor metadata pre-scan so missing trust or visibility aborts before any `SKILL.md` file reads happen.
- Changed visible skill bundle reads from sequential reads to parallel reads, then reconstructs the final candidate list in descriptor order.
- Fixed candidate ordering keys to include `skill_md_path`, using a length-prefixed key format to avoid separator collisions for bundles with the same source/name and different manifest paths.
- Added doc comments for the public `SkillBundleContextSource::new` and `bundle_source` methods.
- Added regression coverage for:
  - missing trust metadata with no reads
  - missing visibility metadata with no reads
  - later descriptor policy failure preventing earlier visible reads
  - same source/name descriptors with different `skill_md_path`
  - list source error mapping
  - visible read `PermissionDenied` mapping
  - `ContentTooLarge` mapping
  - invalid UTF-8 `SKILL.md` parsing failure

Verification run locally:

```text
cargo test -p ironclaw_loop_support
cargo clippy -p ironclaw_loop_support --all-targets -- -D warnings
cargo fmt --check -p ironclaw_loop_support
git diff --check
```

File: crates/ironclaw_loop_support/src/skill_bundle_context_source.rs [hunk 5]
```diff
@@ -0,0 +1,691 @@
...
+                .bundle_source
+                .read_skill_bundle_file(run_context, descriptor.id(), descriptor.skill_md_path())
+                .await
+                .map_err(skill_bundle_source_error_to_context_error)?;
+            let skill_md =
+                String::from_utf8(skill_md).map_err(|_| HostSkillContextBuildError::ParseFailed)?;
+
+            candidates.push(
+                HostSkillContextCandidate::new(skill_md, trust, visibility)
+                    .with_ordering_key(ordering_key),
```
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/skill_bundle_context_source.rs]: `with_ordering_key(ordering_key)` makes the emitted skill snippet order depend on descriptor/source ordering, but prompt construction later re-sorts `instruction_snippets` by `snippet_ref`/`safe_summary` before assigning the `msg:snippet...` ordinal. `ThreadBackedLoopContextPort::instruction_snippet_messages_by_ref` rebuilds refs from `build_skill_instruction_snippets` without applying that prompt sort, so a pair such as `system:zeta` and `user:alpha` can make the prompt cite `alpha` as ordinal 0 while the resolver maps ordinal 0 to `zeta`, causing the requested model-message ref to be unavailable. Please make prompt construction and model-message resolution compute skill snippet ordinals from the same ordering, or avoid descriptor ordering keys until that order is carried through both paths.

File: crates/ironclaw_loop_support/src/skill_bundle_context_source.rs [hunk 6]
```diff
@@ -0,0 +1,691 @@
...
+
+fn descriptor_context_ordering_key(descriptor: &SkillBundleDescriptor) -> String {
+    let (source_kind, name, path) = descriptor.ordering_key();
+    length_prefixed_key_components([source_kind.as_str(), name, path])
+}
+
+fn length_prefixed_key_components<const N: usize>(components: [&str; N]) -> String {
+    let mut key = String::new();
+    for component in components {
+        key.push_str(&component.len().to_string());
```
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/skill_bundle_context_source.rs]: This length-prefixed key is collision-safe, but it is not order-preserving when compared lexicographically by `SkillRunSnapshot`. For example, `13:tenant_shared|...` sorts before `4:user|...`, and `4:user|...` sorts before `6:system|...`, which does not match `SkillBundleDescriptor::ordering_key()` order. If the goal is to preserve descriptor ordering downstream, use an order-preserving encoding such as a zero-padded ordinal assigned after sorting descriptors, or a source-rank prefix plus escaped components.

File: crates/ironclaw_loop_support/src/skill_bundle_context_source.rs [hunk 7]
```diff
@@ -0,0 +1,691 @@
...
+    error: SkillBundleSourceError,
+) -> HostSkillContextBuildError {
+    // Collapse bundle-source internals into the public-safe context error taxonomy:
+    // unavailable, parse/policy failure, budget exhaustion, or internal bug.
+    match error {
+        SkillBundleSourceError::SourceUnavailable
+        | SkillBundleSourceError::BundleNotFound
+        | SkillBundleSourceError::FileNotFound
+        | SkillBundleSourceError::PermissionDenied => HostSkillContextBuildError::SourceUnavailable,
+        SkillBundleSourceError::InvalidBundleId
```
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/skill_bundle_context_source.rs:132]: The new error taxonomy maps several parse-class `SkillBundleSourceError` variants to `HostSkillContextBuildError::ParseFailed`, but the adjacent tests do not exercise `InvalidBundleId`, `InvalidFilePath`, `InvalidSkillBundle`, `BundleUtf8DecodeFailed`, or `ManifestParseFailed`. A targeted test like `adapter_maps_parse_source_errors` over those variants would lock this public error contract down.

File: crates/ironclaw_loop_support/src/skill_bundle_context_source.rs [hunk 8]
```diff
@@ -0,0 +1,691 @@
...
+        for descriptor in &descriptors {
+            let trust = descriptor.trust().cloned();
+            let visibility = descriptor.visibility().copied();
+            let ordering_key = descriptor_context_ordering_key(descriptor);
+
+            if visibility != Some(SkillVisibility::Visible) {
+                // Preserve host policy metadata on unavailable candidates so downstream
+                // snapshot construction keeps one fail-closed validation path.
+                candidates.push(
+                    HostSkillContextCandidate::unavailable(trust, visibility)
```
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/skill_bundle_context_source.rs]: The hidden/denied branch intentionally preserves trust, visibility, and ordering metadata by returning unavailable candidates, but the current test only verifies the final snippet list is empty. That would still pass if this branch silently dropped hidden/denied descriptors. Please add a direct `load_skill_context_candidates` test that asserts hidden and denied candidates are returned with `skill_md == None`, preserved visibility, and no file reads.

**Review by serrrfirat — COMMENTED** [review]
Multi-agent review for PR #3848 at `b3eace586424cf47571cde3d7c125a1fa416fdf0`.

Reviewers completed: security, bugs, performance/concurrency, tests, conventions. I filtered weaker findings that either depended on broader existing trust-model policy or caching design rather than this diff. The actionable findings below focus on ordering correctness and regression coverage for the new adapter.

Summary:
- 2 behavioral findings
- 2 test-coverage findings
- Event: comment

File: crates/ironclaw_loop_support/src/skill_bundle_context_source.rs [hunk 9]
```diff
@@ -0,0 +1,741 @@
...
+    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        formatter
+            .debug_struct("SkillBundleContextSource")
+            .field("bundle_source", &"<SkillBundleSource>")
+            .finish()
+    }
+}
+
+#[async_trait]
+impl<S> HostSkillContextSource for SkillBundleContextSource<S>
```
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/skill_bundle_context_source.rs:58]: The new public `SkillBundleContextSource` is tested directly and through `build_skill_instruction_snippets`, but not through the caller-visible `ThreadBackedLoopContextPort::load_loop_context` path that will actually expose these snippets. The repo testing rule asks for caller-level coverage when a transform/helper sits between computed inputs and a side-effecting flow. Please add a contract test, for example `thread_context_port_builds_skill_instruction_snippets_from_skill_bundle_context_source`, that installs this adapter over a small `SkillBundleSource` test double, calls `load_loop_context`, and asserts visible trusted content appears while installed prompt content stays out of snippets.

File: crates/ironclaw_loop_support/src/skill_bundle_context_source.rs [hunk 10]
```diff
@@ -0,0 +1,741 @@
...
+                // Preserve host policy metadata on unavailable candidates so downstream
+                // snapshot construction keeps one fail-closed validation path.
+                candidates.push(
+                    HostSkillContextCandidate::unavailable(trust, visibility)
+                        .with_ordering_key(ordering_key),
+                );
+                continue;
+            }
+
+            let skill_md = self
```
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/skill_bundle_context_source.rs]: `load_skill_context_candidates` awaits each visible bundle read serially during prompt/context assembly. With the filesystem source this can be up to the bounded bundle scan set, and the trait is async enough that future sources may be slower than local disk. Consider bounded concurrency, preserving descriptor order when collecting results, or caching the per-run skill snapshot before prompt/model materialization.

File: crates/ironclaw_loop_support/src/lib.rs [hunk 11]
```diff
@@ -1167,6 +1171,15 @@ where
     }
 }
 
+fn compare_skill_instruction_snippets(
```
Comment (serrrfirat) [inline crates/ironclaw_loop_support/src/lib.rs]: This duplicates the prompt builder's skill-snippet ordering logic so model-message resolution can recreate ordinal-bearing refs. Today it matches `instruction_bundle.rs` for skill snippets, but the comparator there is private; future prompt-ordering changes can make these refs drift. A shared helper in `ironclaw_turns` for the skill-ref ordering would keep prompt construction and model-message resolution tied to the same rule.

**Review by serrrfirat — COMMENTED** [review]
Multi-agent code review for PR #3848 at `a00991a627a7e6a244d297867a745c626118198e`.

Reviewers completed: security, bugs, performance/concurrency, tests, conventions.

Findings: 3 total
- Medium: 2
- Low: 1
- Clean reviewers: security, bugs

Intent understood: adapt first-party skill bundles into runtime skill-context candidates while preserving trust/visibility enforcement, deterministic ordering, and fail-closed policy behavior.

Verdict: COMMENT. No security or correctness blockers found, but the adapter still needs caller-level regression coverage, and there are two maintainability/performance concerns worth addressing before this becomes the runtime composition path.

**Review by serrrfirat — COMMENTED** [review]
## PR Review Summary
**PR:** nearai/ironclaw#3848 — Adapt skill bundles into Reborn skill context
**Base:** e5ee0b98 → **Head:** 262107cf
**Files changed:** 7 (755 lines in new adapter, 282 in integration tests)

### Statistics
| Category          | Critical | High | Medium | Low | Nit |
|-------------------|----------|------|--------|-----|-----|
| Security          | 0        | 0    | 1      | 0   | 0   |
| Bugs              | 0        | 0    | 0      | 0   | 0   |
| Performance       | 0        | 0    | 1      | 3   | 1   |
| Tests             | 0        | 1    | 2      | 1   | 0   |
| Conventions       | 0        | 0    | 0      | 0   | 0   |
| **Total**         | **0**    | **1**| **4**  | **4**| **1**| 

---

### 🔴 Blocking (High)

**[Tests] FileNotFound/BundleNotFound unmapped in read path** (confidence: 95)
Function `skill_bundle_source_error_to_context_error` exhaustively maps 13 `SkillBundleSourceError` variants; `FileNotFound` and `BundleNotFound` are never exercised by tests. The mock naturally returns `FileNotFound` for missing files, but every visible-bundle test either provides the file or sets an explicit `read_error`. The `BundleNotFound` variant has never appeared in any test at all.
- `adapter_maps_file_not_found_read_error`: visible bundle without SKILL.md in mock → assert `SourceUnavailable`
- `adapter_maps_bundle_not_found`: inject `BundleNotFound` via `.with_read_error(..., BundleNotFound)` → assert `SourceUnavailable`

---

### 🟡 Medium Findings

**[Security] Unbounded visible bundle count** (line ~62)
`load_skill_context_candidates` reads SKILL.md for every visible descriptor via `stream.buffered(8)`, but there is no cap on descriptor count or total aggregate bytes. Per-file size limits exist in `FilesystemSkillBundleSource` but not at the adapter level for descriptor enumeration.

**[Performance] No per-read size cap on SKILL.md content** (line ~88)
`String::from_utf8(skill_md)` allocates the full SKILL.md before any size check. With `buffered(8)`, worst case is 8× the largest SKILL.md in memory simultaneously.

**[Tests] Empty descriptors list returns empty OK** (line ~64)
The happy-path empty list returns `Ok(vec![])` but no test verifies this path. The list-error tests pass `Vec::new()` but that triggers the `list_error` error path. The happy-path empty case is untested.

**[Concurrency] try_collect aborts on first read failure** (line ~74)
A single `FileNotFound` or `PermissionDenied` read short-circuits the entire candidate list via `try_collect`. All validated descriptors for bundles N+1..M are lost. Acceptable for MVP fail-closed semantics, but worth noting.

---

### 🟢 Low/Nit (informational)
- **N+1 snippet reads**: `build_skill_instruction_snippets` called twice per loop iteration (load_loop_context + instruction_snippet_messages_by_ref). Cached result reused only when OnceCell is set.
- **Magic constant**: `MAX_CONCURRENT_SKILL_BUNDLE_CONTEXT_READS = 8` not parameterizable.
- **futures crate overhead**: Full `futures` crate pulled in; only `StreamExt`, `TryStreamExt`, `stream`, and `buffered` used.
- **Conventions**: Clean — no anchorable issues.

---

### ✅ Good
- Fail-closed on missing trust/visibility metadata is well-implemented.
- Deterministic ordering key preserves descriptor order through sorting.
- Test coverage for error mapping is comprehensive (15+ tests covering all listed variants).
- SkillFilePath validates mount escape thoroughly.
- `sort_instruction_snippets_for_prompt` extracted cleanly for reuse.
